Fix `nextSibling` bugs and refactor `updateNodes()` and its fuzzer

kfule opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR fixes `nextSibling`-related bugs in `updateNodes()` and updates updateNodesFuzzer to better cover these cases. It also refactors `updateNodes()` to simplify the logic and improve maintainability.

## Description
<!--- Describe your changes in detail -->
This PR makes the following changes related to `updateNodes()`.
- Fix for bugs related to nextSibling
- Fix for updateNodesFuzzer and addition of test cases
- Refactoring including adjusting variable declaration positions and names
  - This change reduces the bundle size to 8,899 bytes gzipped.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This PR fixes the following issues related to the way `nextSibling` is set, which I noticed while reading the `updateNodes()` code. (The cases below are simplified versions of those found by the fuzzer.)
Additionally, since updateNodesFuzzer also appeared to have problems, I fixed those issues as well and added additional test cases.

case 1 ([flems](https://flems.io/mithril@2.3.8#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgCsEAaEAYwHs0AXGWvKtOGgAgCdLK2BeVgE0rkArjlr4ARpX4BPADpoFTFqxis+ACgDWASnUA+Vlm0lWwLTBmJWWgL57WjpwHpnq2GLYA3NNJhLqFQAHdVZtPR5DYy1WAGpWORAgxNNzS2s7PVd3GE9WHz8wjF8aAAsYdlYaDABzHQUA5jZKAEZQ5BgNRK0WxJ0STu6AJj6BrpAtAGZRwYmAFj6AXUaVLza+ZCct1iDxrRGQftmtBcOx7unD5cU0bPJ2GAw6BSx8B7R+Co1OblM1+rQykosHwUEoNQ0AEF2OwMDJ8GBOMYfjR8ORStB+AA5PxwHT4LAYXZoAysND4XyfLHYGA6LJuTYAcgA0kMAAqM0wsuac1gsyaM1iLVhUGEwcg0Bq3NzCIL8J7+NCvd6fdjfLg0UzIRYA5X0VXq36sVq6t76r4ov4tAFAkFgiHQ2HwxGUZEatEYqDY3H4wnE0nkykwak4OmsbJM5k8rnMgUx9mC4UQQGUMUSkC2Ra2IA))
```js
const root = document.body

const e = (k) => m(k, {key: k})       // element vnode
const p = (k) => m(k + "p", {key: k}) // element vnode (another tag)

const o1 = [e("k1"),e("k2"),e("k3"),e("k4")]
const v1 = [        p("k2"),e("k4"),e("k3")]

// create
m.render(root, v1)
console.log(Array.from(root.childNodes).map(n => n.nodeName)) // [ 'K2P', 'K4', 'K3' ] correct

// update
m.render(root, [])
m.render(root, o1)
m.render(root, v1)
console.log(Array.from(root.childNodes).map(n => n.nodeName)) // [ 'K4', 'K3', 'K2P' ] incorrect
```

case 2 ([flems](https://flems.io/mithril@2.3.8#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgCsEAaEAYwHs0AXGWvKtOGgAgCdLK2BeVgE0rkArjlr4ARpX4BPADpoFTFqxis+ACgDWASnUA+Vlm0lWwLTBmJWWgL57WrAPRPVsMWwBuaaWo0B3CBoACwFKLB0lahUwdVZtPR5DYzkQZFTTc0trOz0XVjB2DABzD1ZvX3jAkMphNkEIhSjmNkoARjjkMA1UrTbUnVNHYdUekC0AJgGAXWaVTw6+ZBgxvoGSFdSMde7eqZAdWcU0LHx2en4Ydg1OblN2yJOzi6ubrhpTBbzXADlffAgzCuNAAQjAwJRztYAMLBaD8Vg0SisQFwYGsCTgyFqCBwVg+NgYVjkOFQBGUWIhXH43wgWzTWxAA))
```js
const root = document.body

const e = (k) => m(k, {key: k})   // element vnode (with dom)
const f = (k) => m("[", {key: k}) // fragment vnode (without dom)

const o1 = [f("k1"),       e("k2")]
const v1 = [e("k1"),e("a"),f("k2")]

m.render(root, o1)
m.render(root, v1) // Node.insertBefore: Child to insert before is not a child of this node
```

case 3 ([flems](https://flems.io/mithril@2.3.8#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgCsEAaEAYwHs0AXGWvKtOGgAgCdLK2BeVgE0rkArjlr4ARpX4BPADpoFTFqxis+ACgDWASnUA+Vlm0lWwLTBmJWWgL57WrAPRPVsMWwBuaaWo0B3CBoACwFKLB0lahUwdVZtPR5DYzkQZFTTc0trOz0XVjB2DABzD1ZvX3jAkMphNkEIhSjmNkoARjjkMA1UrTbUnRJu3oAmAZIYHpAtAGYBgF1mlU8OvmRJ3v6QQeHpue2Jqa0x7cXFNCx8dnp+GHYNTm5TdsiLq5u7h64aUxW81wAcr58BBmHcaAAhGBgSjXawAYWC0H4rBolFYoLg4NYEmhsLUEDgrB8bAwrHISKgKMosRChOJvhAtnmtiAA))
```js
const root = document.body

const e = (k) => m(k, {key: k})   // element vnode (with dom)
const f = (k) => m("[", {key: k}) // fragment vnode (without dom)

const o1 = [f("k1"),f("k2"),e("k3")]
const v1 = [e("k1"),f("k3"),e("k2")]

m.render(root, o1)
m.render(root, v1) // Node.insertBefore: Child to insert before is not a child of this node
```

Since these cases occur only with keyed vnodes and involve very rare patterns - such as fragments that do not correspond to DOM nodes or tag changes that cause node deletions and insertions - the bugs likely did not surface very often.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
`npm run test`
I have also confirmed the above flems case fix.

As for the fuzzer, I’ve confirmed that it is tested with different patterns, and that even after increasing the number of runs, all tests still pass.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] My change requires a documentation update, and I've opened a pull request to update it already:
- [x] I have read https://mithril.js.org/contributing.html.

kfule committed:
Process updates in reverse order of `old` instead of `vnodes`.

Because the DOM has not yet been arranged in `vnodes` order, `nextSibling` can point to the wrong DOM node. Since `insertBefore` uses `nextSibling`, nodes may be inserted in the wrong position when deletions and insertions occur. This commit fixes that by performing updates based on `old`, which reflects the actual DOM order.

kfule committed:
Adjust the scanning range of `getNextSibling()`.

Previously, `getNextSibling()` scanned all the way to the end of the array. If the array contained vnodes that did not correspond to DOM nodes, or if newly inserted vnodes were present within the scan range, the scan could overrun and capture an incorrect nextSibling. This commit fixes that by limiting the scan range appropriately.

kfule committed:
updateNodesFuzzer: Properly capture test data and comment out test cases that do not account for the heuristics in `updateNodes()`.

Previously, due to a bug that caused the same data to be reused, the fuzzer was not exercising the code adequately. It also included test cases that ignored the heuristics in `updateNodes()`, which had a very low probability of being triggered but, once they were, produced a large burst of false-positive test failures from that same data.
The commented-out sections are test cases for the number of append and insert operations related to the longest increasing subsequence algorithm, so it should be safe to leave them commented out for now.

kfule committed:
updateNodesFuzzer: Added test cases for tag changes (involving deletions and insertions), as well as cases with fragments that do not correspond to DOM nodes.

kfule committed:
[refactor]updateNodes: Move index increment logic into the unkeyed block.

By specializing this logic for the unkeyed path, this change reduces the number of conditional branches and also shrinks the bundle size.

kfule committed:
[refactor]updateNodes: Move map creation outside the loop and inline/optimize `getKeyMap()`.

In v1, `getKeyMap()` was called inside the loop, but in v2 this is no longer necessary. Since key normalization has been improved in v2, much of `getKeyMap()`’s logic has become redundant, so this change inlines and simplifies it to reduce overhead.

dead-claudia commented on render/render.js:
```diff
@@ -347,17 +347,19 @@ module.exports = function() {
-							old[oldIndex] = null
+					var originalNextSibling = nextSibling, pos = 2147483647, matched = 0
+					var oldIndices = new Array(end - start + 1)
+					for (var i = 0; i < oldIndices.length; i++) oldIndices[i] = -1
```
Can we do this instead? Or is it slower?

```suggestion
					var oldIndices = new Int32Array(end - start + 1).fill(-1)
```

kfule replied:
It seems like a good idea.
Since this code path isn't tested with `npm run perf`, I'll take a quick look using dbmonster or some other benchmark.

kfule replied:
I tried running dbmonster, but I couldn’t tell if there was any noticeable difference. So I changed it to `new Array(end - start + 1).fill(-1)`, which reduces the bundle size.

dead-claudia commented on render/render.js:
```diff
@@ -497,7 +488,7 @@ module.exports = function() {
 		var il = lisTemp.length = a.length
 		for (var i = 0; i < il; i++) lisTemp[i] = a[i]
-		for (var i = 0; i < il; ++i) {
+		for (var i = 0; i < il; i++) {
```
Can you revert this change? I'd rather keep the Git blame cleaner

kfule replied:
reverted it

dead-claudia commented on render/tests/test-updateNodesFuzzer.js:
```diff
@@ -25,18 +25,44 @@ o.spec("updateNodes keyed list Fuzzer", function() {
+			const id = i++
+			o(id + ": " + test.list.join() + " -> " + test.updated.join(), function() {
 				render(root, test.list.map(function(x){return m(x, {key: x})}))
 				addSpies(root)
 				render(root, test.updated.map(function(x){return m(x, {key: x})}))
 
-				if (root.appendChild.callCount + root.insertBefore.callCount !== test.expected.creations + test.expected.moves) console.log(test, {aC: root.appendChild.callCount, iB: root.insertBefore.callCount}, [].map.call(root.childNodes, function(n){return n.nodeName.toLowerCase()}))
+				// FIXME: This does not take into account the "swaps and list reversals" heuristic in updateNodes().
+				// if (root.appendChild.callCount + root.insertBefore.callCount !== test.expected.creations + test.expected.moves) console.log(test, {aC: root.appendChild.callCount, iB: root.insertBefore.callCount}, [].map.call(root.childNodes, function(n){return n.nodeName.toLowerCase()}))
```
Could you elaborate on what the specific issue is with this?

This feels a bit too generic. Given the nature of the issue it's checking, and the fact it's just auditing for perf opportunities, it might take years to get fixed, assuming it gets fixed at all.

kfule replied:
You're absolutely right. I'd like to leave a more detailed comment.

dead-claudia commented on render/tests/test-updateNodesFuzzer.js:
```diff
@@ -25,18 +25,44 @@ o.spec("updateNodes keyed list Fuzzer", function() {
 
-				o(root.appendChild.callCount + root.insertBefore.callCount).equals(test.expected.creations + test.expected.moves)("moves")
+				// ditto
+				// o(root.appendChild.callCount + root.insertBefore.callCount).equals(test.expected.creations + test.expected.moves)("moves")
```
See above comment

dead-claudia reviewed (changes requested):
In addition to the other comments, can you also put in some deterministic tests to ensure we don't regress in at least the cases you listed?

If they break, I'd rather us see it right away. I don't want to risk the bug unexpectedly reappearing making me roll for initiative after a bad release. 😅 (This may be before your time here, but we got burned *hard* in a v0.2 release thanks to missing test coverage.)

kfule committed:
[refactor]updateNodes: Adjust variable declaration positions and names.

These changes aim to improve code readability and slightly reduce the bundle size.

kfule committed:
updateNodesFuzzer: Add detailed comments for FIXME

kfule committed:
Add deterministic tests for nextSibling handling

kfule commented:
@dead-claudia Thank you for your review comments.
I’ve made the corresponding changes for each of your points. I’ve also added deterministic tests.
Could you please take another look?
